Split out separate qiskit-transpiler crate

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit creates a new qiskit-transpiler crate. As we move towards exposing more rust functionality directly via the C API having more explicit interface boundaries in the rust code is important for keeping things organized and managing dependencies (both external and internal). We also are building towards a case where the core functionality exists without a dependency on python, so more explicitly splitting functionality makes it clearer where the expected boundaries are. The ideal end state is where we can separate the API of qiskit-transpiler (and the existing qiskit-circuit) to be pure rust and then move the python interface components to only live in the accelerate or pyext crate. But that is a longer term vision and this is only one step towards that goal.

The transpiler functionality in Rust has grown to the point where it is both sufficiently self contained and also large enough to warrant its own compilation unit. Especially because the transpiler is the next big frontier in the C API, so having it be standalone will make it simpler to expose via cext.

This PR isn't a 100% clean separation yet though. Ideally qiskit-transpiler would not depend on qiskit-accelerate. However this PR still has qiskit-transpiler dependency on qiskit-accelerate. This is for 2 reasons primarily. The first is there are synthesis and quantum info pieces that are still in qiskit-accelerate which several passes depend. The next step after this PR is to migrate those into a standalone crate for synthesis and/or quantum info.  The second component is the nlayout module. This should live in the transpiler crate, but was not migrated as part of this PR. This will be done in a follow-up as ideally we should move PhysicalQubit and VirtualQubit to qiskit-circuit, and the NLayout struct should move to transpiler. This was deferred as part of this commit to minimize the changes done at once.

Related to this there are two modules which are moved to the new transpiler crate which should ideally live in qiskit-accelerate, twirling and circuit_duration. However, they depend on the target and a transpiler pass, and until we can reverse the dependency between accelerate and transpiler we had to move these.

### Details and comments

mtreinish committed:
Split out separate qiskit-transpiler crate

This commit creates a new qiskit-transpiler crate. As we move towards
exposing more rust functionality directly via the C API having more
explicit interface boundaries in the rust code is important for keeping
things organized and managing dependencies (both external and internal).
We also are building towards a case where the core functionality exists
without a dependency on python, so more explicitly splitting
functionality makes it clearer where the expected boundaries are. The
ideal end state is where we can separate the API of qiskit-transpiler
(and the existing qiskit-circuit) to be pure rust and then move the
python interface components to only live in the accelerate or pyext
crate. But that is a longer term vision and this is only one step
towards that goal.

The transpiler functionality in Rust has grown to the point where it is
both sufficiently self contained and also large enough to warrant its
own compilation unit. Especially because the transpiler is the next big
frontier in the C API, so having it be standalone will make it simpler
to expose via cext.

This PR isn't a 100% clean separation yet though. Ideally
qiskit-transpiler would not depend on qiskit-accelerate. However this PR
still has qiskit-transpiler dependency on qiskit-accelerate. This is for
2 reasons primarily. The first is there are synthesis and quantum info
pieces that are still in qiskit-accelerate which several passes depend.
The next step after this PR is to migrate those into a standalone crate
for synthesis and/or quantum info.  The second component is the nlayout
module. This should live in the transpiler crate, but was not migrated
as part of this PR. This will be done in a follow-up as ideally we
should move PhysicalQubit and VirtualQubit to qiskit-circuit, and the
NLayout struct should move to transpiler. This was deferred as part of
this commit to minimize the changes done at once.

Related to this there are two modules which are moved to the new transpiler
crate which should ideally live in qiskit-accelerate, twirling and
circuit_duration. However, they depend on the target and a transpiler pass,
and until we can reverse the dependency between accelerate and transpiler
we had to move these.

mtreinish commented:
An open question here is what to do with the equivalence library implementation. In Python space it's part of the circuit module. Although I'm not sure the exact rationale for why it lives there. So ideally it would probably live in `qiskit-circuit` if we were going to match that. Although the only use case in rust space is the the basis translator and it seems more logically placed in the transpiler module. I think we want to move it out of accelerate so we can reverse the dependency between qiskit-accelerate and qiskit-transpiler.

If putting it in `qiskit-transpiler` makes more sense I'll do that in this PR, but if it's qiskit-circuit then I'll make a new PR on-top of #14287 for that.

raynelfss commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -42,20 +42,20 @@ use qiskit_circuit::{
 use qiskit_circuit::{Clbit, Qubit};
```
I had this file as part of a `basis` folder with the idea that more transpiler passes that worked with basis would eventually come along, is that no longer the plan?

mtreinish replied:
The structure for the python space tree didn't really fit here. In Python it's defined around the passes that nominally go into a stage, but also not really because there isn't a perfect mapping. The structure that's there now is mostly historical because it predates the definition of the formal stages which is why there isn't an exact mapping (there is no `init` directory for example). For example, we run `ConsolidateBlocks` in `init`, `optimization`, and optionally in `translation`,  but it's in the `optimization/` directory. Changing this in Python is more disruptive than it is useful, while we say it's only stable at the documented import path it's a needless breakage for people ignoring that to just rename or shuffle around things without a good reason.

In this specific case I decided it was more trouble than it's worth because we're adding another path component where there is nothing else there yet. If having a `basis` directory/module makes sense in the future when we have > 1 pass in rust that we categorize that way. But right now there are only 22 passes and that kind of nested organizational structure feels unnecessary at this point.

However, this doesn't have to be the final organizational structure we can organize things into subdmodules as we see fit in later PRs without any disruption (except for potential merge conflicts in github, git itself can deal with code moving well enough). I wouldn't be opposed to a follow up commit trying to organize things into directories based on nominal stage that uses a pass.

raynelfss commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -42,20 +42,20 @@ use qiskit_circuit::{
+use crate::target::QargsRef;
+use crate::target::Target;
+use qiskit_accelerate::equivalence::EquivalenceLibrary;
+use qiskit_accelerate::nlayout::PhysicalQubit;
```
Might be better to just join these imports since they come from the same module.
```suggestion
use crate::target::exceptions::TranspilerError;
use crate::target::{Qargs, QargsRef, Target};
use qiskit_accelerate::equivalence::EquivalenceLibrary;
use qiskit_accelerate::nlayout::PhysicalQubit;
```

mtreinish replied:
I didn't want to change any of the code outside what was needed here to minimize the diff for review and potentially bisect in the future. This applies to all the proposed inline suggestions around consolidating imports. It's easy enough to do this in a follow up PR, but for something like this where I'm moving about 10k lines of code around I didn't want to introduce more differences between the files in their original location and here than was necessary.

raynelfss commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -27,12 +27,15 @@ use qiskit_circuit::Qubit;
+    TwoQubitBasisDecomposer, TwoQubitControlledUDecomposer,
+};
+
+use crate::target::Qargs;
+use crate::target::Target;
```
```suggestion
use crate::target::{Qargs, Target};
```

raynelfss commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -704,7 +704,7 @@ impl EquivalenceLibrary {
     }
```
Is there a world where the `EquivalenceLibrary` could be part of the `qiskit_transpiler` crate? As far as I'm aware, it seems to be only used with transpiler passes and not much more after that.

mtreinish replied:
See: https://github.com/Qiskit/qiskit/pull/14286#issuecomment-2848126067 it's an option. It's more where we want it to live.

raynelfss replied:
**Here are my two cents:** I think it makes sense to move it to `transpiler` because it doesn't make sense for it to be anywhere else functionally speaking.

Functionally speaking, this structure doesn't add any features to the circuits but instead uses them. If we were to compare it to another existing library within the circuit submodule, the circuit library, things are very different. Whereas the circuit library contains all of the gates +  example circuits that could be added onto a circuit, extending what the user can do when building a one; The equivalence library maps gates onto equivalent circuits, which doesn't really add functionality to the circuit itself. 

If a user was to manually change the gates in a circuit with its equivalents using the EquivalenceLibrary, they would just end up with the same circuit, but with different building blocks. Again not adding functionality but manually transpiling it.

Seeing that the purpose of the structure is to change a gate based on whatever equivalence is available (mostly dictated by the basis gates specified by the target), which is only done during transpilation, it makes sense for this to be in the `transpiler` crate rather than circuit or accelerate (since it isn't really an accelerator).

I wonder if the rest would agree.

mtreinish replied:
I moved it into this crate in: https://github.com/Qiskit/qiskit/pull/14286/commits/17f33c5e736b07e463e6aca1ac248a2593001a2c (along with the commutation checker).

raynelfss commented on crates/transpiler/src/passes/high_level_synthesis.rs:
```diff
@@ -39,14 +39,14 @@ use qiskit_circuit::Clbit;
+use qiskit_accelerate::euler_one_qubit_decomposer::angles_from_unitary;
+use qiskit_accelerate::euler_one_qubit_decomposer::EulerBasis;
+use qiskit_accelerate::nlayout::PhysicalQubit;
+use qiskit_accelerate::two_qubit_decompose::TwoQubitBasisDecomposer;
```
```suggestion
use crate::target::exceptions::TranspilerError;
use crate::target::{Qargs, Target};
use qiskit_accelerate::equivalence::EquivalenceLibrary;
use qiskit_accelerate::euler_one_qubit_decomposer::{angles_from_unitary, EulerBasis};
use qiskit_accelerate::nlayout::PhysicalQubit;
use qiskit_accelerate::two_qubit_decompose::TwoQubitBasisDecomposer;
```

raynelfss commented on crates/transpiler/src/passes/mod.rs:
```diff
@@ -0,0 +1,65 @@
+// This code is part of Qiskit.
```
Should the passes mod include a readme too? I understand it isn't a crate on its own but since it has all of the transpiler passes, I think we should consider it (Unless it goes against the documentation convention here).

mtreinish replied:
I can update the docs to contain more details. Realistically I'd probably prefer us using more rustdoc here and putting the details in module level docs rather than in the readme, but I was sparse on the docs on both accounts.in this PR so I'll add more details.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14286/commits/08a6544e2620bde8c77445a5290894f0ae8f7830

raynelfss commented on crates/transpiler/src/passes/star_prerouting.rs:
```diff
@@ -24,23 +24,23 @@ type Nodes = (usize, Vec<VirtualQubit>, HashSet<usize>, bool);
-use crate::sabre::BlockResult;
-use crate::sabre::NodeBlockResults;
-use crate::sabre::SabreResult;
+use super::sabre::sabre_dag::SabreDAG;
+use super::sabre::swap_map::SwapMap;
+use super::sabre::BlockResult;
+use super::sabre::NodeBlockResults;
+use super::sabre::SabreResult;
```
```suggestion
use super::sabre::sabre_dag::{SabreDAG, SwapMap};
use super::sabre::{BlockResult, NodeBlockResults, SabreResult};
```

raynelfss commented on crates/transpiler/src/passes/star_prerouting.rs:
```diff
@@ -24,23 +24,23 @@ type Nodes = (usize, Vec<VirtualQubit>, HashSet<usize>, bool);
 use hashbrown::HashSet;
 use numpy::IntoPyArray;
 use pyo3::prelude::*;
+use qiskit_accelerate::nlayout::PhysicalQubit;
+use qiskit_accelerate::nlayout::VirtualQubit;
```
```suggestion
use qiskit_accelerate::nlayout::{PhysicalQubit, VirtualQubit};
```

mtreinish committed:
Merge remote-tracking branch 'origin/main' into qiskit-transpiler-crate

raynelfss commented on crates/README.md:
```diff
@@ -8,10 +8,14 @@ The intention is that (much) longer term, we might be wanting to expose more of
 * `qiskit-pyext` is the only crate that actually builds a Python C extension library.
```
Should the `transpiler` crate be exposed as something other than `qiskit._accelerate` to Python? I feel like we are way beyond accelerators at this point and should have a distinction of which rust crates are being used. Maybe we could have `qiskit._accelerate`, `qiskit._circuit`, and `qiskit._transpiler`. Though I understand it might be confusing since it re-uses the same names, it should be considered as rust has become a bigger part of our codebase.

This would be outside of the scope of this PR but I wanted to bring it up here since we're already organizing.

mtreinish replied:
No I don't think it should. This was something we actively did in the past and it caused more headaches than it was worth. Having the single Python entrypoint is desirable for a lot of reasons (you can look at the docs and commits around when pyext was added for the context there).

mtreinish committed:
Add thiserror to qiskit-transpiler crate post-rebase

raynelfss reviewed (commented):
Thank you for doing this, I'm happy we're separating the transpiler aspects of Qiskit from the accelerators, this is something we should have done long ago. I have only a couple questions and comments.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into qiskit-transpiler-crate

mtreinish committed:
Update imports in new tests

mtreinish committed:
Merge remote-tracking branch 'origin/main' into qiskit-transpiler-crate

mtreinish committed:
Expand docs around submodules

mtreinish committed:
Move CommutationChecker and Equivalence library to transpiler too

This commit moves the commutation checker and the equivalence library to
the new transpiler crate too. These structures while potentially more
general are in practice only ever used by the transpiler so in the
logical grouping of functionality it makes the most sense as part of the
transpiler crate. If we decide this doesn't make sense in the future we
can move either to qiskit-circuit or a different crate in the future.
But for right now this feels like the best match and also helps remove
the dependency of qiskit-transpiler on qiskit-accelerate.

raynelfss reviewed (approved):
LGTM!
